### ISSUE TITLE
FSDP optimizer state dict APIs

### ISSUE THREAD

**Issue by zhaojuanmao** [pytorch/pytorch#73884]
### 🚀 The feature, motivation and pitch

FSDP flatten parameters and shard these flatten parameters among ranks.  After that, these sharded flatten parameters are passed to optimizers. Optimizers will optimize for these sharded flatten parameters locally in the training loop. 

When saving original unflatten parameters' states in optimizer, we need APIs to help all_gather the sharded flatten parameters, and then unflatten them back to original ones. Also, for non-tensor states or local states, FSDP needs to make sure they are saved properly as well for the optimizers. 

before calling optimizer.load_state_dict() for FSDP model, FSDP needs APIs to load these saved original unflatten parameters' states, flatten them, shard them and then pass the sharded states to optimizer.load_state_dict(). 

The implementation will be similar to 'gather_full_optim_state_dict()' and 'get_shard_from_optim_state_dict()' in FairScale FSDP. In default, for parameters' states, it can output states with "indexes" as keys. But based on users' needs, we may provide one more option to output states with  "parameter names" as keys. 

An example to save optimizer states: 
   fsdp_model = fsdp(model, old_process_group)
  optimizer = torch.adam(fsdp_model.parameters())
   # training loop
   .....
   full_optim_state_dict = fsdp_model.gather_full_optim_state_dict(optimizer)
   torch.save(full_optim_state_dict)

An example to load optimizer states with new process group and new world_size:
      fsdp_model = fsdp(model, new_process_group)
      optimizer = torch.adam(fsdp_model.parameters())
      optimizer.load_state_dict(fsdp_model.get_shard_from_optim_state_dict(full_optim_state_dict))
       # training loop
       .....

**Comment by awgu** [comment]
Initial version is landed; see the following stack of diffs: https://github.com/pytorch/pytorch/pull/74215

**Comment by awgu** [comment]
> How would `FSDP.full_optim_state_dict(model, optim)` handle a optimizer created with multiple networks, like `torch.optim.AdamW(itertools.chain( net1.parameters(), net2.parameters(), ... ),...)`?

I believe that the FSDP API would not support that. Would it be possible to break up the optimizer into one `AdamW` per root module (i.e. per `net1`, `net2`, ...)?

cc: @fegin
If I understand correctly, we _could_ support this by permitting the `model: nn.Module` argument to also have something like type `Sequence[nn.Module]` and handle this internally (by additionally looping over _all_ passed-in modules). However, we have not had motivation to do that until this question. If breaking up the optimizer works, then we probably prefer that to keep the API simpler.

**Comment by fegin** [comment]
@awgu, yup, it is possible to do that but also agree breaking up the optimizer or grouping nets into one module is the preferred way. `optim_state_dict` is in the refactor stage and we would like to simplify the logic, adding `Sequence[nn.Module]` support would not help at the moment. Before `optim_state_dict` reaches a more stable stage, I prefer not to adding this support.
